pkg/sync/rpc/server.go

```diff
@@ -445,16 +447,77 @@ func (s *SyncAgentServer) FilesSync(ctx context.Context, req *enginerpc.FilesSyn
 		}
 	}()
 
+	if req.LocalSync != nil {
+		err := s.fileSyncLocal(ctx, req)
+		if err == nil {
+			return &emptypb.Empty{}, nil
+		}
+
+		logrus.WithError(err).Warn("Falling back to remote sync")
+	}
+
+	return &emptypb.Empty{}, s.fileSyncRemote(ctx, req)
+}
+
+func (s *SyncAgentServer) fileSyncLocal(ctx context.Context, req *enginerpc.FilesSyncRequest) error {
+	var targetPaths []string
+	var err error
+
+	log := logrus.WithFields(logrus.Fields{
+		"source": req.LocalSync.Source,
+		"target": req.LocalSync.Target,
+	})
+
+	log.Info("Syncing files locally")
+
+	// Defer function to handle cleanup of files if an error occurs
+	defer func() {
+		if err == nil {
+			log.Info("Done syncing files locally")
+		} else {
+			log.WithError(err).Warn("Failed to sync files locally, reverting changes")
+
+			for _, targetPath := range targetPaths {
+				if removeErr := os.Remove(targetPath); removeErr != nil && removeErr != os.ErrNotExist {
+					log.WithError(removeErr).Warnf("Failed to remove file %v", targetPath)
+				}
+			}
+		}
+	}()
+
+	for _, info := range req.SyncFileInfoList {
+		sourcePath := filepath.Join("/host", req.LocalSync.Source, info.FromFileName)
+		targetPath := filepath.Join("/host", req.LocalSync.Target, info.ToFileName)
+		targetPaths = append(targetPaths, targetPath)
+
+		log.Tracef("Copying file %v to %v", sourcePath, targetPath)
+
+		err = lhio.CopyFile(sourcePath, targetPath, true)
+		if err != nil {
+			return err
+		}
```

Comment:
`lhio.CopyFile` is based on `io.Copy` ([codes](https://github.com/longhorn/go-common-libs/blob/d78642cff5e1e810d919ba3db2698a4acff59a99/io/file.go#L137-L140)). `io.Copy` doesn't support sparse file copy.

Reply:
https://github.com/longhorn/sparse-tools/pull/104